File: src/main/java/architecture/ego_equipment/datagen/i18n/ZhCn.java
package architecture.ego_equipment.datagen.i18n;

import architecture.ego_equipment.core.EGOEquipment;
import architecture.goldenboughs_lib.datagen.i18n.DatagenI18n;
import net.minecraft.data.PackOutput;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.item.Item;
import net.neoforged.fml.loading.FMLEnvironment;
import org.jetbrains.annotations.ApiStatus;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

@ApiStatus.Internal
public final class ZhCn extends DatagenI18n {
	private static final Map<Supplier<? extends Item>, String> ITEMS = new HashMap<>();
	private static final Map<Supplier<? extends SoundEvent>, String> SOUND_EVENT = new HashMap<>();
	private static final Map<String, String> MAP = new HashMap<>();
	private static final Map<Supplier<? extends EntityType<?>>, String> ENTITY = new HashMap<>();

	public ZhCn(final PackOutput output) {
		super(output, EGOEquipment.ID, "zh_cn");
	}

	public static void addI18nText(String zhCn, String key) {
		if (!FMLEnvironment.production) {
			MAP.put(key, zhCn);
		}
	}

	public static void addI18nItemText(String zhName, Supplier<? extends Item> deferredItem) {
		if (!FMLEnvironment.production) {
			ITEMS.put(deferredItem, zhName);
		}
	}

	public static void addI18nSoundEventText(String zhName, Supplier<? extends SoundEvent> supplier) {
		if (!FMLEnvironment.production) {
			SOUND_EVENT.put(supplier, zhName);
		}
	}

	public static void addI18nEntityTypeText(String zhName, Supplier<? extends EntityType<?>> supplier) {
		if (!FMLEnvironment.production) {
			ENTITY.put(supplier, zhName);
		}
	}

	@Override
	protected void addTranslations() {
		addPackDescription(EGOEquipment.ID, "E.G.O.装备");
		addItemList(ITEMS);
		addEntityList(ENTITY);
		addSoundEventList(SOUND_EVENT);
		MAP.forEach(this::add);
	}
}

File: src/main/kotlin/architecture/ego_equipment/datagen/i18n/ZhCn.kt
package architecture.ego_equipment.datagen.i18n

import architecture.ego_equipment.core.EGOEquipment
import architecture.goldenboughs_lib.datagen.i18n.DatagenI18n
import net.minecraft.data.PackOutput
import net.minecraft.sounds.SoundEvent
import net.minecraft.world.entity.EntityType
import net.minecraft.world.item.Item
import net.neoforged.fml.loading.FMLEnvironment
import org.jetbrains.annotations.ApiStatus

@ApiStatus.Internal
class ZhCn(output: PackOutput) : DatagenI18n(output, EGOEquipment.ID, "zh_cn") {

	companion object {
		private val ITEMS: MutableMap<() -> Item, String> = HashMap()
		private val SOUND_EVENT: MutableMap<() -> SoundEvent, String> = HashMap()
		private val MAP: MutableMap<String, String> = HashMap()
		private val ENTITY: MutableMap<() -> EntityType<*>, String> = HashMap()

		@JvmStatic
		fun addI18nText(zhCn: String, key: String) {
			if (!FMLEnvironment.production) {
				MAP[key] = zhCn
			}
		}

		@JvmStatic
		fun addI18nItemText(zhName: String, deferredItem: () -> Item) {
			if (!FMLEnvironment.production) {
				ITEMS[deferredItem] = zhName
			}
		}

		@JvmStatic
		fun addI18nSoundEventText(zhName: String, supplier: () -> SoundEvent) {
			if (!FMLEnvironment.production) {
				SOUND_EVENT[supplier] = zhName
			}
		}

		@JvmStatic
		fun addI18nEntityTypeText(zhName: String, supplier: () -> EntityType<*>) {
			if (!FMLEnvironment.production) {
				ENTITY[supplier] = zhName
			}
		}
	}

	override fun addTranslations() {
		addPackDescription(EGOEquipment.ID, "E.G.O.装备")
		addItemList(ITEMS)
		addEntityList(ENTITY)
		addSoundEventList(SOUND_EVENT)
		MAP.forEach { (key, value) -> add(key, value) }
	}
}
